game-core/src/main/java/games/strategy/triplea/ai/pro/util/MoveBatcher.java

```diff
@@ -0,0 +1,164 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * MoveBatcher implements an algorithm that batches AI moves together, reducing the number of
+ * distinct moves the AI does (and thus the visual delay) by combining moves together.
+ */
+public class MoveBatcher {
+  private static class Move {
+    private final ArrayList<Unit> units;
+    private final Route route;
+    private final HashMap<Unit, Unit> unitsToTransports;
+
+    private Move(
+        final ArrayList<Unit> units,
+        final Route route,
+        final HashMap<Unit, Unit> unitsToTransports) {
+      this.units = units;
+      this.route = route;
+      this.unitsToTransports = unitsToTransports;
+    }
+
+    Move(final ArrayList<Unit> units, final Route route) {
+      this(units, route, null);
+    }
+
+    Move(final Unit unit, final Route route, final Unit transportToLoad) {
+      this(
+          mutableSingletonList(unit),
+          route,
+          new HashMap<Unit, Unit>(Map.of(unit, transportToLoad)));
+    }
+
+    private static ArrayList<Unit> mutableSingletonList(final Unit unit) {
+      return new ArrayList<>(Collections.singletonList(unit));
+    }
+
+    boolean isTransportLoad() {
+      return this.unitsToTransports != null;
+    }
+
+    boolean canMergeWith(final Move other) {
+      return other.isTransportLoad() == isTransportLoad() && route.equals(other.route);
+    }
+
+    boolean mergeWith(final Move other) {
+      if (canMergeWith(other)) {
+        // Merge units and transports.
+        units.addAll(other.units);
+        if (isTransportLoad()) {
+          unitsToTransports.putAll(other.unitsToTransports);
+        }
+        return true;
+      }
+      return false;
+    }
+
+    void addTo(
+        final List<Collection<Unit>> moveUnits,
+        final List<Route> moveRoutes,
+        final List<Map<Unit, Unit>> unitsToTransports) {
+      moveUnits.add(units);
+      moveRoutes.add(route);
+      unitsToTransports.add(this.unitsToTransports);
+    }
+  }
+
+  private final ArrayList<ArrayList<Move>> moveSequences = new ArrayList<ArrayList<Move>>();
+
+  /**
+   * Starts a new sequence. This must be called before calling any add*() methods. A sequence
+   * indicates a logical dependency relationship between the moves in that sequence.
+   */
+  public void newSequence() {
+    moveSequences.add(new ArrayList<Move>());
+  }
+
+  /**
+   * Adds a transport load move. Ownership of the params is transferred to this object.
+   *
+   * @param unit The unit to move.
+   * @param route The route to move on.
+   * @param transport The transport to load.
+   */
+  public void addTransportLoad(final Unit unit, final Route route, final Unit transport) {
+    addMove(new Move(unit, route, transport));
+  }
+
+  /**
+   * Adds a move for the specified units. Ownership of the params is transferred to this object.
```

Comment:
Would you clarify what is meant by: " Ownership of the params is transferred to this object." ?

Reply:
Also, it's possible this all becomes moot after the refactoring to not use parallel lists.

Because after that refactoring, we may be passing Move objects around instead of these list params, which I either will allow us to be both efficient and nice as an API (i.e. whereas ArrayList doesn't have a swap() method that can be done efficiently, if we have a wrapper object like Move(), we may be able to do this - or if we make Move objects immutable, then this is also moot.)

Reply:
Nice APIs are worth quite a bit, impossible to be used incorrectly, means no bugs and fast development as you don't have to study any API you use, you can just use it. For maintenance, it also means modularity and isolation, to reason about the state of this object, you have to know about the callers, and from the caller perspective, you have to know about the implementation of this object and its callers.

With respect to runtime cost, anything order nano-second is too fast to worry about, premature optimization. I ran an experiment once to check the differing runtimes over different JDKs, from 1.2 to 8. In Java 8, new object creation can be faster than calling a no-op method, things have gotten really fast and well optimized.  For appreciation of how fast a ns is, I very much like this blog post: https://blog.codinghorror.com/the-infinite-space-between-words/

This is Java after-all, IMO avoiding in/out or out parameters is good, copying in parameters and returning the value is a standard'ish approach, nobody is going to fault you for the list copy.